extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,150 @@
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_RETRIES = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG = 'amp-ad-batch-manager';
+
+/** @const {Object} A map of AmpAdBatchManager objects, one for each value of data-url on the page */
+let imageadBatchManagers = null;
+
+/** @const {Object} A map of arrays of imagead Elements, keyed by data-url */
+let imageadElements = null;
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} elem The element which is to be batched
+  * @param {string} url The URL identifying which batch
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(elem, url) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (imageadBatchManagers === null) {
+    // Scan the page to get a list of all image ads on this page, indexed by url.
+    // For performance reasons, we only do the scan once, the first time that a batch manager is constructed.
+    imageadBatchManagers = {};
+    imageadElements = {};
+    const elements = document.querySelectorAll('amp-ad[type=imagead]');
+    for (let ind = 0; ind < elements.length; ind++) {
+      const el = elements[ind];
+      const url = el.getAttribute('data-url');
+      if (!(url in imageadElements)) {
+        imageadElements[url] = [];
+      }
+      imageadElements[url].push(el);
+    }
+  }
+  if (!(url in imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    elem.batchMaster(true);
+    imageadBatchManagers[url] = new AmpAdBatchManager(url,
+        imageadElements[url]);
+  }
+  return imageadBatchManagers[url];
+}
+
+export class AmpAdBatchManager {
+
+  /**
+   * @param {string} url The base URL of the ad server
+   * @param {Array} elements An array of imagead Element objects to be batched together */
+  constructor(url, elements) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = url;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullUrl_ = this.url_;
+
+    /** @private {number} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.retries_ = AD_BATCH_RETRIES;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Get a list of slots that are in use for this URL
+    const slots = elements.map(element => {
+      const slotId = element.getAttribute('data-slot');
+      // Use a slot id of 0 if the element didn't provide one.
+      return slotId ? slotId : 0;
```

Comment:
Carrying on the conversation from https://github.com/ampproject/amphtml/pull/5751#discussion_r87081984, I don't think this is the right default. If people want to display the same ad in multiple slots, let's make them request it specifically.